File: autopilot/lib/paralib/path.cpp
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include "path.h"

#ifdef UNIT_TEST
#define PRINTF printf
#else
#include <Arduino.h>
#define PRINTF Serial.printf
#endif

/**
 * Construct a new path object.
 */
Path *new_path(const char *name, uint8_t segment_count, Segment *segments[]) {
  if (segment_count <= 0) {
    PRINTF("Error: new_path segment_count %d\n", segment_count);
    return NULL;
  }
  // C style malloc for variable length struct
  Path *path = (Path*) malloc(sizeof(Path) + segment_count * sizeof(Segment*));
  if (!path) {
    PRINTF("Error: new_path malloc failed %p\n", path);
    return NULL;
  }
  path->name = name;
  path->start = segment_start(segments[0]);
  path->end = segment_end(segments[segment_count - 1]);
  // Copy segments to struct
  path->segment_count = segment_count;
  for (int i = 0; i < segment_count; i++) {
    path->segments[i] = segments[i];
  }
  return path;
}

/**
 * Fly the path a given distance, and then free the original.
 * This is helpful for flying a path in place.
 * Consumes the passed path.
 */
Path *path_fly_free(Path *path, double distance) {
  if (!path) return NULL;
  if (!(distance > 0)) {
    PRINTF("path fly distance must be positive %f\n", distance);
    return path;
  }

  double flown = 0;
  int i = 0;
  Segment *trimmed[path->segment_count + 1];
  for (; i < path->segment_count - 1; i++) {
    const double len = segment_length(path->segments[i]);
    if (distance < flown + len) {
      // End is within segment
      break;
    } else {
      // Copy whole segment
      trimmed[i] = path->segments[i];
      flown += len;
    }
  }

  const double remaining = distance - flown;
  if (remaining > 0) {
    // Fly last segment
    Path *last = segment_fly(path->segments[i], distance - flown);
    // Free last segment
    delete path->segments[i];
    // Concat last segment(s)
    trimmed[i++] = last->segments[0];
    // A turn can become a turn plus a line
    if (last->segment_count == 2) {
      trimmed[i++] = last->segments[1];
    }
    // Free last path, segments have been consumed
    free(last);
  } else if (remaining < 0) {
    PRINTF("segment_fly distance must be positive %f - %f < 0\n", distance, flown);
  }

  // Construct new path
  Path *flight = new_path(path->name, i, trimmed);

  // Free remaining segments of previous path
  for (; i < path->segment_count; i++) {
    delete path->segments[i];
  }
  free(path);

  return flight;
}

double path_length(Path *path) {
  double len = 0;
  for (int i = 0; i < path->segment_count; i++) {
    len += segment_length(path->segments[i]);
  }
  return len;
}

ParaControls path_controls(Path *path) {
  Segment *segment = path->segments[0];
  if (segment->segment_type == 'T') {
    Turn *turn = (Turn *) segment;
    return turn_controls(turn);
  } else {
    // Straight
    ParaControls ctrl = {};
    return ctrl;
  }
}

void free_path(Path *path) {
  if (path) {
    // Free segments
    for (int i = 0; i < path->segment_count; i++) {
      delete path->segments[i];
    }
    free(path);
  }
}

File: simulator/ts/geo/path.ts
import { Point, PointV, TogglePosition } from "../dtypes"
import { SegmentLine } from "./segment-line"
import { SegmentTurn } from "./segment-turn"

export type Segment = SegmentLine | SegmentTurn

export class Path {
  readonly name: string
  readonly start: PointV
  readonly end: PointV
  readonly segments: Segment[]

  constructor(name: string, ...segments: Segment[]) {
    this.name = name
    this.segments = segments
    this.start = segments[0].start
    this.end = segments[segments.length - 1].end
  }

  public controls(): TogglePosition {
    return this.segments[0].controls()
  }

  public length(): number {
    return this.segments
      .map((s) => s.length())
      .reduce((a, b) => a + b, 0)
  }

  public render(): Point[] {
    // Flatten segments
    return ([] as Point[]).concat(...this.segments.map((s) => s.render()))
  }

  /**
   * Fly a given distance along the path.
   * If the end of the path is reached, extend a straight line out to infinity.
   */
  public fly(distance: number): Path {
    if (!(distance > 0)) {
      console.error("path fly distance must be positive", distance)
      return this
    }
    const trimmed: Segment[] = []
    let flown = 0
    let i = 0
    for (; i < this.segments.length - 1; i++) {
      const segment = this.segments[i]
      const segmentLength = segment.length()
      if (distance < flown + segmentLength) {
        // End is within segment
        break
      } else {
        trimmed.push(segment)
        flown += segmentLength
      }
    }
    const remaining = distance - flown
    if (remaining > 0) {
      // Fly last segment
      const last = this.segments[i].fly(distance - flown)
      trimmed.push(...last.segments)
    } else if (remaining < 0) {
      console.error("remaining distance must be positive", distance, flown)
    }
    return new Path(this.name, ...trimmed)
  }

}
